actor_library/src/payment/escrow.rs

```diff
@@ -0,0 +1,194 @@
+use anyhow::anyhow;
+use cid::Cid;
+use fil_actors_runtime::runtime::Runtime;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_hamt::BytesKey;
+use fvm_shared::{
+    address::Address,
+    bigint::Zero,
+    econ::TokenAmount,
+    event::{Entry, Flags},
+    IPLD_RAW,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::utils::map;
+use crate::{
+    context::Context,
+    ownable::{DefaultOwnable, Ownable},
+};
+
+const DEPOSITED_EVENT_KEY: &str = "deposited_event";
+const WITHDRAWN_EVENT_KEY: &str = "withdrawn_event";
+
+#[derive(Serialize, Deserialize)]
+pub struct DepositedEvent {
+    pub payee: Address,
+    pub amount: TokenAmount,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct WithdrawnEvent {
+    pub payee: Address,
+    pub payment: TokenAmount,
+}
+
+pub trait Escrow {
+    fn deposits_of<BS: Blockstore>(&self, store: &BS, payee: Address) -> TokenAmount;
+
+    fn deposit<BS: Blockstore>(&mut self, store: &BS, payee: Address) -> anyhow::Result<()>;
+
+    fn withdraw(&mut self, rt: &impl Runtime, payee: Address) -> anyhow::Result<()>;
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct DefaultEscrow {
+    ownable: DefaultOwnable,
+
+    // Mapping(Address -> TokenAmount)
+    deposits: Cid,
+}
+
+impl Context for DefaultEscrow {}
+
+impl DefaultEscrow {
+    pub fn new<BS: Blockstore>(store: &BS) -> anyhow::Result<DefaultEscrow> {
+        let empty_deposits_map = map::make_empty_map::<_, TokenAmount>(store).flush()?;
+
+        Ok(Self {
+            ownable: DefaultOwnable::new(),
+            deposits: empty_deposits_map,
+        })
+    }
+}
+
+impl Escrow for DefaultEscrow {
+    fn deposits_of<BS: Blockstore>(&self, store: &BS, payee: Address) -> TokenAmount {
+        let deposits = map::make_map_with_root::<_, TokenAmount>(&(self.deposits), store).unwrap();
+        let key = payee.to_bytes();
```

Comment:
Resolve address to make sure that the same address format is always used.

```suggestion
        let payee = match resolve_address(&payee) {
            Some(payee) => payee,
            None => {
                return Err(anyhow!(
                    "ERC20: cannot resolve address in initial balance table"
                ))
            }
        };
        let key = payee.to_bytes();
```

Reply:
Fixed [here](https://github.com/Ethernal-Tech/filecoin-actor-library/pull/14/commits/ec51a6b30d4ee1edf85a4878022d07b2b35b8741#diff-a4a6fe0b07d4cafcdcdc97f0bcd54b2e1190edcfed3ffc12199489732766787bR68)